### ISSUE TITLE
Maybe worth automate one entry to have it really fully automated...

### ISSUE THREAD

**Issue by ZumbiAzul** [FrancescoSaverioZuppichini/Pytorch-how-and-when-to-use-Module-Sequential-ModuleList-and-ModuleDict#10]
This is just a suggestion. There is one entry which would depend on the data itself. Maybe it is worth calculating it automatically as well? Any ideas on how to do it?

```
class MyCNNClassifier(nn.Module):
    def __init__(self, in_c, enc_sizes, dec_sizes,  n_classes, activation='relu'):
        super().__init__()
        self.enc_sizes = [in_c, *enc_sizes]
        self.dec_sizes = [32 * 28 * 28, *dec_sizes] # here: 32 * 28 *28

        self.encoder = MyEncoder(self.enc_sizes, activation=activation)
        
        self.decoder = MyDecoder(dec_sizes, n_classes)
        
    def forward(self, x):
        x = self.encoder(x)
        
        x = x.flatten(1) # flat
        
        x = self.decoder(x)
        
        return x
```

**Comment by FrancescoSaverioZuppichini** [comment]
Hi @ZumbiAzul  in modern cnn architecture (> resnet) the features of the last layer are first feed into a an adaptive pooling that removes the channels and creates a (W,H) tensor that can be flatted. You can do something like:

```python
class MyDecoder(nn.Module):
    def __init__(self, in_features, n_classes):
        super().__init__()
        self.avg = nn.AdaptiveAvgPool2d((1, 1))
        self.decoder = nn.Linear(in_features, n_classes)

    def forward(self, x):
        x = self.avg(x)
        x = x.view(x.size(0), -1)
        x = self.decoder(x)
        return x
```

```python
class MyCNNClassifier(nn.Module):
    def __init__(self, in_c, enc_sizes, dec_sizes,  n_classes, activation='relu'):
        super().__init__()
        self.enc_sizes = [in_c, *enc_sizes]

        self.encoder = MyEncoder(self.enc_sizes, activation=activation)
        self.decoder = MyDecoder(enc_sizes[-1], n_classes)
        
    def forward(self, x):
        x = self.encoder(x)
               
        x = self.decoder(x)
        
        return x
```

**Comment by ZumbiAzul** [comment]
That's a good point, @FrancescoSaverioZuppichini , thanks!
Also, thanks for the useful article.

**Comment by ZumbiAzul** [comment]
Hi @FrancescoSaverioZuppichini . First, happy New Year!
I am currently trying to use your tutorial together with you suggestion to use `AdaptiveAvgPool2D`, however, I still have an error of dimensions mismatch. I posted my question on PyTorch [here](https://discuss.pytorch.org/t/using-adaptiveavgpool2d-runtimeerror-mat1-dim-1-must-match-mat2-dim-0/108110). Could you please take a look in case you could spot the issue?
Thanks in advance!
